### PR TITLE
Add authorization for POST events

### PR DESCRIPTION
**Pull request by acn-sbuad**
## Related Issue(s)
- #183

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Altinn.Platform.Events.Authorization
+{
+    /// <summary>
+    /// Authorization handler for the <see cref="PublishScopeOrAccessTokenRequirement"/> class.
+    /// </summary>
+    public class PublishScopeOrAccessTokenHandler : AuthorizationHandler<PublishScopeOrAccessTokenRequirement>
```
Comment (SandGrainOne) [inline src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:14]: I would prefer these two policies to be separate. Is that causing issues?

Comment (SandGrainOne) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:14]: Read a bit more about it. We must implement this as one policy, but it should be possible for one Policy to use two handlers. Might be something to look into?

Comment (acn-sbuad) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:14]: Spent quite some time before I landed on this solution. Couldn't find anything else feasible to handle the OR condition. So I would need a concrete example or a commit to the PR to handle the implementation differently. 

That being said, there are two handlers being used. We are using the handler made available to the authorization service. Lines 33 and 41 call existing policies registered in Program.cs

Comment (acn-sbuad) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:14]: I would suggest we keep it as is. Keeps us from adding an extra class and introducing more boilerplate code while also making it less obvious how the access is granted. One would need to check all handlers to see which requirement they are related to instead of it being 1:1 between requirement and handler.

File: src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            _httpContextAccessor = httpContextAccessor;
+        }
+
+        /// <summary>
+        /// Handles verification of requirements. Publish scope requirement or AccesToken requirement must be satisfied.
+        /// </summary>
+        protected override async Task<object> HandleRequirementAsync(AuthorizationHandlerContext context, PublishScopeOrAccessTokenRequirement requirement)
+        {
+            var authorizationService = _httpContextAccessor.HttpContext.RequestServices.GetService<IAuthorizationService>();
```
Comment (SandGrainOne) [inline src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:31]: Could we try to depend on IAuthorizationService directly instead of going through the HttpContextAccessor?

Comment (acn-sbuad) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:31]: Nope. it results in a circular dependency unfortunately..

Comment (acn-sbuad) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:31]: "A circular dependency was detected for the service of type 'Microsoft.AspNetCore.Authorization.IAuthorizationService'.
Microsoft.AspNetCore.Authorization.IAuthorizationService(Microsoft.AspNetCore.Authorization.DefaultAuthorizationService) -> Microsoft.AspNetCore.Authorization.IAuthorizationHandlerProvider(Microsoft.AspNetCore.Authorization.DefaultAuthorizationHandlerProvider) -> System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Authorization.IAuthorizationHandler> -> Microsoft.AspNetCore.Authorization.IAuthorizationHandler(Altinn.Platform.Events.Authorization.AccessTokenHandler) -> Microsoft.AspNetCore.Authorization.IAuthorizationService"

Comment (SandGrainOne) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:31]: hmm, okay. As we talked about. Fixing this probably requires changes in the AccessToken library. Maybe we should look into that?

Comment (SandGrainOne) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:31]: Actually, what if the new requirement inherit from AccessTokenRequirement. That way AccessTokenHandler would accept the new requirement as input. Then make a new Handler for the scope check that use the same requirement.

Could that work?

Comment (SandGrainOne) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:31]: I see one possible issue with my last idea. A requirement can't inherit from both ScopeAccessRequirement and AccessTokenRequirement. At least one of them would  need to be an interface.

File: test/Altinn.Platform.Events.Tests/TestingControllers/EventsControllerTests.cs [hunk 3]
```diff
@@ -491,20 +492,87 @@ public async Task Post_AuthorizedRequest_EventIsRegistered()
...
+                Mock<IEventsService> eventMock = new();
+                eventMock.Setup(em => em.RegisterNew(It.IsAny<CloudEvent>()))
+                          .ReturnsAsync(Guid.NewGuid().ToString());
+
+                Mock<IAuthorization> authorizationMock = new();
+                authorizationMock.Setup(a => a.AuthorizePublishEvent(It.IsAny<CloudEvent>()))
+                            .ReturnsAsync(false);
+
+                HttpClient client = GetTestClient(eventMock.Object, authorizationMock.Object, true);
+                client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", PrincipalUtil.GetOrgToken("ttd"));
```
Comment (SandGrainOne) [inline test/Altinn.Platform.Events.Tests/TestingControllers/EventsControllerTests.cs:540]: Should this have been a user token instead of an org token? Ref the test name.

Comment (acn-sbuad) [reply test/Altinn.Platform.Events.Tests/TestingControllers/EventsControllerTests.cs:540]: Renamed test.

File: src/Events/Extensions/UriExtensions.cs [hunk 4]
```diff
@@ -12,6 +12,8 @@ namespace Altinn.Platform.Events.Extensions
     /// </summary>
     public static class UriExtensions
     {
+        private static string _urnPattern = @"\burn:[a-z0-9][a-z0-9-]{0,31}:[a-z0-9()+,\-.:=@;$_!*'%/?#]+";
```
Comment (elsand) [inline src/Events/Extensions/UriExtensions.cs:15]: This will allow strings like `urn:1:::::""""` which we probably don't want.

```suggestion
        private static string _urnPattern = @"^urn:[a-z][a-z0-9-]{0,31}(:[a-z][a-z0-9()+,.\-=@;$_!*'%#]{0,99}){1,10}\/?$";
```
* Add `$` to disallow arbitrary trailing characters
* Replace `\b` with `^` to disallow whitespace at the beginning
* Require all "parts" to begin with a letter (ie. disallow "urn:1:2")
* Limit each "part" beyond the first to max 100 characters
* Disallow empty "parts" (ie. `urn:altinn:foo::::`)
* Require 1 to 10 parts beyond the first

Comment (acn-sbuad) [reply src/Events/Extensions/UriExtensions.cs]: Change seems to be breaking some of the unit tests. Will have a look

Comment (acn-sbuad) [reply src/Events/Extensions/UriExtensions.cs]: `urn:isbn:0451450523` is a valid urn ref https://en.wikipedia.org/wiki/Uniform_Resource_Name.
Rolling back to existing regex for now. Consider reworking the regex so existing unit tests run green @elsand ?

Comment (b-brodie-digdir) [reply src/Events/Extensions/UriExtensions.cs]: Could `System.Uri.IsWellFormedUriString`  with UriKind.Absolute and a check for scheme `urn` be an alternative?

https://learn.microsoft.com/en-us/dotnet/api/system.uri.iswellformeduristring?view=net-7.0

File: src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs (hunk 2 above)
Comment (b-brodie-digdir) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:31]: @SandGrainOne did you find a way to use multiple handlers, like we looked at on Wednesday?

File: src/Events/Extensions/UriExtensions.cs (hunk 4 above)
Comment (acn-sbuad) [reply src/Events/Extensions/UriExtensions.cs]: Did a quick test and `urn:altinn:foo:::::` passes the check for System.Uri.IsWellFormedUriString. I believe we verify that it is a well formed uri before we reach this point as well.

Comment (acn-sbuad) [reply src/Events/Extensions/UriExtensions.cs]: That being said.. if Microsoft deems it a well formed uri, we might be happy enough with that? Easier than maintaining the regex string at least.

Comment (elsand) [reply src/Events/Extensions/UriExtensions.cs]: `^urn:[a-z0-9-]{1,30}(:[a-z0-9()+,.\-=@;$_!*'%#\/]{1,100}){1,10}$` will allow numerics as the first (or sole) character in both NID and NSS parts. This also includes `/` as a valid NSS character, which seems reasonable (as we allow it as a trailing character)

We should perhaps consider restricting our support to a smaller subset of that standard ref wikipedia. I'm not sure it will ever make sense here to support percent-encoded arbitrary bytes in the NSS part, or query-components and fragments. I'm not too confident allowing eg. NUL bytes (even encoded) won't have any potential security implications down the line.

`^urn:[a-z0-9-]{1,30}(:[a-z0-9()+,.\-;$_!\/]{1,100}){1,10}$` disallows `% ? #` while still passing the unit tests and support a reasonably wide range of subdivisions in the NSS-es.

On the other hand, if we _are_ to support query-components and fragments, I think we should attempt to validate that those are well-formed (perhaps by explicitly calling `IsWellFormedUriString`)

Comment (b-brodie-digdir) [reply src/Events/Extensions/UriExtensions.cs]: If `System.Uri.IsWellFormedUriString` is _that_ lax, then for sure, being more restrictive sounds like a good idea. Could maybe add some test coverage for empty URN segments, to make sure we don't have unintended wildcard matching in the middle of a URN path.

File: src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs (hunk 2 above)
Comment (SandGrainOne) [reply src/Events/Authorization/PublishScopeOrAccessTokenHandler.cs:31]: @b-brodie-digdir yes, we're currently testing changes in AccessTokenHandler and I've made a PR to update ScopeAccessHandler in the same way.

**Review by SandGrainOne — APPROVED** [review]

File: src/Events/Extensions/UriExtensions.cs (hunk 4 above)
Comment (acn-sbuad) [reply src/Events/Extensions/UriExtensions.cs]: I think we can follow up on this at a later point.. Regardless the URN they choose for the resource will have to pass checks when being registered in the resource registry as well. We will sync with them on the validation.
